python/ray/data/_internal/lazy_block_list.py

```diff
@@ -566,25 +621,60 @@ def _num_computed(self) -> int:
         return i
 
 
-def _execute_read_task(
+def _execute_read_task_nosplit(
     i: int,
     task: ReadTask,
     context: DatasetContext,
     stats_uuid: str,
     stats_actor: ray.actor.ActorHandle,
-) -> Tuple[MaybeBlockPartition, BlockPartitionMetadata]:
+) -> Tuple[Block, BlockMetadata]:
     DatasetContext._set_current(context)
     stats = BlockExecStats.builder()
 
-    # Execute the task.
-    block = task()
+    # Execute the task. Expect only one block returned when dynamic block splitting is
+    # not enabled.
+    blocks = list(task())
+    assert len(blocks) == 1
+    block = blocks[0]
 
     metadata = task.get_metadata()
-    if context.block_splitting_enabled:
-        metadata.exec_stats = stats.build()
-    else:
+    metadata = BlockAccessor.for_block(block).get_metadata(
+        input_files=metadata.input_files, exec_stats=stats.build()
+    )
+    stats_actor.record_task.remote(stats_uuid, i, [metadata])
+    return block, metadata
+
+
+def _execute_read_task_split(
+    i: int,
+    task: ReadTask,
+    context: DatasetContext,
+    stats_uuid: str,
+    stats_actor: ray.actor.ActorHandle,
+) -> Iterable[Union[Block, List[BlockMetadata]]]:
+    """Execute read task with dynamic block splitting.
+
+    Returns an Iterable of blocks followed by their metadata.
```

Comment:
Can we return an Iterable[Tuple[Block, BlockMetadata]] which is more readable and consistent with non-split task?

Reply:
Currently Ray core generator API only supports yield one variable at a time. So we have to yield all blocks metadata as a list, as the last element.